README.md
Byteking agent
==============
[![NPM version](http://img.shields.io/npm/v/byteking-agent.svg)](https://www.npmjs.com/package/byteking-agent)

The new generation monitoring and analytics service agent for ambitious applications.

Description
-----------
Nodejs agent provides transfer client data to ByteKing servers. Separated receiving service ensure the best network package 
loss protection. It can be used as both independent service (soon) or part of your application.

Installation
------------
It can be easily installed by npm.
```bash
npm install byteking-agent
```

Include installed package to your script and initialise with configuration file to run agent. Please see configuration example in `/example/custom_example.json`
```js
const ByteKingAgent = require('byteking-agent');
ByteKingAgent.run({
                      udp_server: {
                          port: 40001
                      },
                      transmitter: {
                          url:'http://127.0.0.1:8080'
                      },
                      debug: false
                  });
```

Then
>> node bk.js \
--debug true \
--udp_server:port 40000 \
--udp_server:address 127.0.0.1
--transmitter:url http://127.0.0.1:8090 \
--transmitter:data_send_force 1000 \
--transmitter:data_send_interval 10000 \


Byteking agent get metric data from the web client(s).
Official clients:
- PHP https://github.com/startbase/byteking-php

Contribution
----
Thank you for interesting in Byteking. If you intend to make Byteking better you are fantastic.
There are some ways to help to improve this product:
- Just use it and submit issues if you find ones
- Submit pull request

TODO
----
- Test coverage


CHANGELOG.md
v0.2.4 (2018-01-12) 
================================================================================
- Removed nconf library

v0.2 (2017-08-01)
================================================================================
- Added argv params for run
- Remove trash
- Global refactoring..

v0.1.4 (2017-07-19)
================================================================================
- Added socket.io server, which means agent is able to receive messages by web socket

v0.1.3 (2017-06-23)
================================================================================
- Added readme and changelog

app.js
const path = require('path');
const fs = require('fs');
const dgram = require('dgram');
const server = dgram.createSocket('udp4');
const WebSocketClient = require(path.join(__dirname, './ws-client'));

const CONFIG_FILE = 'config.json';

function App() {
    this.data_queue = [];
    this.socket = null;

    this.initUpdServer = function(params) {
        server.on('error', function(err) {
            console.log('server error: ', err);
            server.close();
        });

        server.on('message', (message) => {
            this.data_queue.push(message.toString());
        });

        server.on('listening', () => {
            const address = server.address();
            console.log('server listening ', address);
        });
        server.bind(params['port']);
    };

    this.initWSTransmitter = function(params) {
        "use strict";
        let url = (params['use_ssl'] !== undefined && params['use_ssl']) ? 'wss' : 'ws';
        url += '://' + params['remote_address'] + ':' + params['remote_port'];
        this.socket = new WebSocketClient(url, params['reconnect_time_interval']);
    };

    this.init = function() {
        "use strict";
        const raw_config = fs.readFileSync(path.join(__dirname, CONFIG_FILE), 'utf8');
        const config = JSON.parse(raw_config);

        this.initUpdServer(config['udp_server']);
        this.initWSTransmitter(config['web_socket_client']);

        setInterval(() => {
            if(!this.socket.isOpen()) {
                return null;
            }
            //@todo переддавать только один api_key для массива
            let stack_length = this.data_queue.length;
            let stack_data = this.data_queue.splice(0, stack_length);

            this.socket.send(JSON.stringify(stack_data));
        }, config['data_transfer_interval']);
    }
}

(new App()).init();


lib/ws-client.js
const WebSocket = require('ws');

function WebSocketClient(url, reconnect_interval) {
    this.reconnect_interval = reconnect_interval;
    this.url = url;

    this.open();
}

WebSocketClient.prototype.open = function() {
    this.instance = new WebSocket(this.url);
    this.instance.on('close',(e) => {
        if(e !== 1000) {
            this.reconnect();
        }
    });
    this.instance.on('error',(e) => {
        console.log('Error code: ' + e.code);
    });

    this.instance.on('connect', () => {
        console.log('Connect established: ' + this.url);
    });
};
WebSocketClient.prototype.send = function(data,option) {
    if(this.isOpen()) {
        this.instance.send(data, option);
    }
};
WebSocketClient.prototype.isOpen = function() {
    return this.instance.readyState === WebSocket.OPEN;
};
WebSocketClient.prototype.reconnect = function() {
    console.log(`Reconnecting in ${this.reconnect_interval}ms`);
    setTimeout(() => {
        this.open();
    }, this.reconnect_interval);
};

module.exports = WebSocketClient;

byteking-agent.js
const path = require('path');
const fs = require('fs');
const dgram = require('dgram');
const server = dgram.createSocket('udp4');
const WebSocketClient = require(path.join(__dirname, './ws-client'));

module.exports = function ByteKingAgent() {
    this.config_dir = path.join(__dirname, 'config.json');
    this.data_queue = [];
    this.socket = null;

    this.initUpdServer = function(params) {
        server.on('error', function(err) {
            console.log('server error: ', err);
            server.close();
        });

        server.on('message', (message) => {
            this.data_queue.push(message.toString());
        });

        server.on('listening', () => {
            const address = server.address();
            console.log('server listening ', address);
        });
        server.bind(params['port']);
    };

    this.initWSTransmitter = function(params) {
        "use strict";
        let url = (params['use_ssl'] !== undefined && params['use_ssl']) ? 'wss' : 'ws';
        url += '://' + params['remote_address'] + ':' + params['remote_port'];
        this.socket = new WebSocketClient(url, params['reconnect_time_interval']);
    };

    this.init = function(config_dir) {
        "use strict";
        config_dir = config_dir || this.config_dir;
        const raw_config = fs.readFileSync(config_dir, 'utf8');
        const config = JSON.parse(raw_config);

        this.initUpdServer(config['udp_server']);
        this.initWSTransmitter(config['web_socket_client']);

        setInterval(() => {
            if(!this.socket.isOpen()) {
                return null;
            }
            //@todo переддавать только один api_key для массива
            let stack_length = this.data_queue.length;
            let stack_data = this.data_queue.splice(0, stack_length);

            this.socket.send(JSON.stringify(stack_data));
        }, config['data_transfer_interval']);
    }
};


index.js
'use strict';

const dgram = require('dgram');
const WebSocket = require('ws');
const md5 = require('md5');
const merge = require('deepmerge');

class ByteKingAgent {
    /**
     * Creates a ByteKingAgent instance.
     */

    constructor() {

        this.config = {
            udp_server: {
                port: 40004,
                address: '0.0.0.0'
            },
            transmitter: {
                url:'',
                host:'',
                port:'',
                use_ssl:false,
                reconnect_time:5000,
                data_send_interval:5000,
                data_send_force:100
            },
            debug:false

        };
    };

    log() {
        if (this.config.debug) {
            console.log(arguments);
        }
    };

    initUdpServer() {
        this.log('init udp server');

        if (!BK.config.udp_server.port) {
            console.error('Error! Please check config, use "--udp_server:port 4000" for example');
            return;
        }


        const server = dgram.createSocket('udp4');
        server.on('error', (err) => {
            this.log('UDP Server error: ', err);
            server.close();
        });

        server.on('message', (message) => {
            this.addMessage(message.toString());
        });

        server.on('listening', () => {
            this.log('UDP Server listening: ', server.address());
        });

        server.bind({port: BK.config.udp_server.port, address: BK.config.udp_server.address});
    };

    initTransmitter() {
        if (this._init_transmitter) {
            return;
        }

        this._init_transmitter = true;

        this.log('init transmitter');
        if (!BK.config.transmitter.url) {
            console.error('Error! Please check config, use "--transmitter:url ws://127.0.0.1:8090" for example');
            return;
        }

        let url = BK.config.transmitter.url;

        this._transmitter = new WebSocket(url);

        this._transmitter.on('open', data => {
            BK._init_transmitter = false;
        });

        this._transmitter.on('message', data => {
            this.log(data);
        });

        this._transmitter.on('error', data => {
            setTimeout(() => {
                BK._init_transmitter = false;
                BK.initTransmitter();
            }, BK.config.transmitter.reconnect_time);
        });

    };

    addMessage(message) {
        if (message.indexOf("_") === 0) {
            this.multiMessage(message);
            return;
        }

        this._data_queue.push(message);
        this.log(this._data_queue.length, this._data_send_force);
        if (this._data_queue.length >= this._data_send_force) {
            clearInterval(this._interval);
            this._send();
            this.initInterval();
        }
    };

    multiMessage(message) {
        let result_size = 0;
        let cur_item = 0;
        let hash = '';
        let msg_hash = '';
        let msg = '';

        if (message.indexOf("_p_") === 0) {
            let slice = 3;
            result_size = message.substring(slice, slice + this.part_block_size);
            slice += this.part_block_size + 1;
            cur_item = message.substring(slice, slice + this.part_block_size);
            slice += this.part_block_size + 1;
            hash = message.substring(slice, slice + 32);
            slice += 32 + 1;
            msg_hash = message.substring(slice, slice + 32);
            slice += 32 + 1;
            msg = message.substring(slice);
        }

        if (!this.multi_msgs.has(hash)) {
            this.multi_msgs.set(hash, new Map());
        }

        let part = this.multi_msgs.get(hash);
        part.set(parseInt(cur_item), msg);

        if (part.size === parseInt(result_size)) {
            if (this.multiMessageCompare(part, msg_hash)) {
                this.multi_msgs.delete(hash);
                return;
            } else {
                setTimeout(() => {
                    this.multiMessageCompare(part, msg_hash);
                    this.multi_msgs.delete(hash);
                }, 1000);
            }
        }
    };

    multiMessageCompare(part, msg_hash) {
        let result = '';
        let parts = [...part.entries()].sort();
        for (let i in parts) {
            if (parts.hasOwnProperty(i)) {
                result += parts[i][1];
            }
        }

        if (md5(result) === msg_hash) {
            this.log('multi success send', msg_hash);
            this.addMessage(result);
            return true;
        }
        return false;
    };

    run(config) {
        if (config) {
            BK.config = merge(BK.config, config);
        }

        this.log('byteking-agent process pid: ', process.pid);


        this._data_send_force = BK.config.transmitter.data_send_force;

        this._run_interval = true;
        this._interval = null;

        this._data_queue = [];
        this._transmitter = null;
        this._init_transmitter = false;
        this.debug = BK.config.debug;
        this.multi_msgs = new Map();

        this.part_block_size = 3;


        this.initUdpServer();
        this.initTransmitter();
        this.initInterval();
        return this;
    };

    initInterval() {
        this._interval = setInterval(() => {
            if (this._run_interval) {
                this._send();
            }
        }, BK.config.transmitter.data_send_interval);
    };

    _send() {
        this.log('send');
        if (!this._transmitter) {
            return;
        }

        if (this._transmitter.readyState !== this._transmitter.OPEN) {
            this.initTransmitter();
            return;
        }

        let data = this._data_queue;
        this._data_queue = [];

        try {
            this._transmitter.send(JSON.stringify(data));
            this.log(data.length);
        } catch (e) {
            this.log("Error send", e.code, e.message);
            this._data_queue.push(...data);
            this.initTransmitter();
            return;
        }

    };
}

let BK = new ByteKingAgent();
module.exports = BK;